proposals/progress_output.md

```diff
@@ -0,0 +1,184 @@
+Note that the build would pause on `Finalizing...` until all layers are finished pushing.
+
+## Goal
+
+To display fine-grained progress feedback during the build, but not clutter the log messages.
+
+## Design considerations
+
+Since Jib builds/pulls and pushes each layer independently, synchronized progress feedback (completion over all layers, such as a progress bar) could potentially limit the concurrency of the build process - but this might be negligible.
+
+The available information for each layer include:
+
+- type (classes, resources, dependencies, base image layer, etc.)
+- digest
+- size
+
+The non-available information for each layer include:
+
+- sizes of other layers (eg. other layers may not have finished building)
+- number of other layers (eg. base image layers have not been resolved before starting to push an application layer)
+
+Therefore, the actions we can perform for progress feedback include:
+
+- monitor the number of bytes sent (and percentage completion)
+
+And the actions we cannot perform include:
+
+- know ahead of time how many layers will be pulled/pushed
+
+## Proposal
+
+Display an overall progress bar along with the tasks currently being executed.
+
+### Example
+
+```
+[=====================            ] 60% complete
+> Pushing classes layer
+> Pulling base image layer 50501d3b88f7
+> Pushing dependencies layer
+> Pushing base image layer 8b106a18283f
+```
+
+The currently executing tasks are displayed below the overall progress bar.
+
+*Note that this would replace the logs messages that are currently outputted as those may corrupt the progress bar display.*
+
+### Summary
+
+The implementation consists of two parts:
+
+1. Emit progress events from the builder steps.
+1. Monitor progress events and display to console.
+
+### Emit progress events
+
+There are a few issues to address for the progress event design:
+
+- the total amount of work is not known beforehand, so there is no static max progress known at any time
+- the builder steps are asynchronous and therefore progress events should not share a single progress state
+- rather, the progress event receiver should be the only potentially stateful entity
+
+These issues can be resolved with a *decentralized allocation tree*.
+
+#### Decentralized allocation tree (DAT)
+
+Each node in the DAT is immutable and initialized with a count of allocation units. These allocation units are to be claimed by progress made on that node or child nodes. Each child node claims 1 allocation unit of the parent node, meaning that completion of all allocation units on that child node means completion of 1 allocation unit on the parent node.
+
+##### Example
+
+Consider a root node (`Node A`) with 10 allocation units. A child node (`Node B`) is added with 3 allocation units. Each of these allocation units represents `10% x 30% = 3%` of the entire DAT progress. Completion of all the 3 child node allocation units means that the root node completed 1 allocation unit, or `10%`.
```

Comment:
Just a nit that I didn't really point out the other day: I think it is `10% x 33.333% = 3.3333%` instead of `10% x 30% = 3%`.

Reply:
Fixed and chose 2.5% to keep it non-repeating.

Reply:
Fixed and chose 2.5% to keep it non-repeating.